[CS] Ensure DeclContext is switched for if/switch expressions

hamishknight opened the pull request:
Similar to multi-statement closures, we need to make sure we change the DeclContext of the ConstraintSystem when solving a conjunction for an if/switch expression. This is unfortunately needed for cases within single-expression closures since they can't switch the system's DeclContext (as they're solved together with the rest of the system).

Resolves #79395

hamishknight commented:
@swift-ci please test

hamishknight commented:
@swift-ci please test source compatibility

hamishknight commented:
@swift-ci please test source compatibility Debug

xedin reviewed (approved):
I think we should also do the same for Tap expressions

hamishknight committed:
[CS] Ensure DeclContext is switched for if/switch expressions

Similar to multi-statement closures, we need to
make sure we change the DeclContext of the
ConstraintSystem when solving a conjunction for an
if/switch expression. This is unfortunately needed
for cases within single-expression closures since
they can't switch the system's DeclContext (as
they're solved together with the rest of the system).

hamishknight commented:
> I think we should also do the same for Tap expressions

Heh, I was thinking about that, but couldn't think of a case that wouldn't also be problematic for single-expression closures in general. But yeah, no harm in doing it regardless :)
